src/pip/_internal/index.py

```diff
@@ -1061,7 +1061,7 @@ def _clean_link(url):
         path = urllib_request.pathname2url(
             urllib_request.url2pathname(result.path))
     else:
-        path = urllib_parse.quote(urllib_parse.unquote(result.path))
+        path = urllib_parse.quote(urllib_parse.unquote(result.path), safe="/@")
```

Comment:
I would add a comment above this line saying that `@` needs to be preserved so that the revision string can be parsed from VCS urls.

Reply:
Done